apiserver/facades/client/application/deployrepository.go

```diff
@@ -131,26 +138,60 @@ func (api *DeployFromRepositoryAPI) DeployFromRepository(arg params.DeployFromRe
 		NumUnits:          dt.numUnits,
 		Placement:         dt.placement,
 		Constraints:       dt.constraints,
-		Resources:         dt.resources,
+		Resources:         pendingIDs,
 	})
+
 	if err != nil {
+		// Remove the pending resources that are added before the AddApplication is called
+		removeResourcesErr := api.state.RemovePendingResources(dt.applicationName, pendingIDs)
+		if removeResourcesErr != nil {
+			logger.Errorf("unable to remove pending resources for %q", dt.applicationName)
+		}
 		return params.DeployFromRepositoryInfo{}, nil, []error{errors.Trace(err)}
 	}
 
-	// Last step, add pending resources.
-	pendingResourceUploads, errs := addPendingResources()
-
 	return info, pendingResourceUploads, errs
 }
 
 // addPendingResource adds a pending resource doc for all resources to be
 // added when deploying the charm. PendingResourceUpload is only returned
 // for local resources which will require the client to upload the
 // resource once DeployFromRepository returns. All resources will be
-// processed. Errors are not terminal.
-// TODO: determine necessary args.
-func addPendingResources() ([]*params.PendingResourceUpload, []error) {
-	return nil, nil
+// processed. Errors are not terminal. It also returns the name to pendingIDs
+// map that's needed by the AddApplication.
+func (api *DeployFromRepositoryAPI) addPendingResources(appName string, uploads map[string]string, res map[string]resource.Meta) (map[string]string, []*params.PendingResourceUpload, []error) {
+	var pendingUploadIDs []*params.PendingResourceUpload
+	var errs []error
+	pendingIDs := make(map[string]string)
+
+	for name, meta := range res {
+		r := resource.Resource{
+			Meta: meta,
+		}
+		if _, ok := uploads[name]; ok {
+			r.Origin = resource.OriginUpload
+		} else {
+			r.Origin = resource.OriginStore
+		}
+
+		pID, err := api.state.AddPendingResource(appName, r)
+		if err != nil {
+			logger.Warningf("Unable to add pending resource %v for application %v", name, appName)
+			errs = append(errs, err)
+		} else {
+			pendingIDs[name] = pID
+			if _, ok := uploads[name]; ok {
+				pendingUploadIDs = append(pendingUploadIDs, &params.PendingResourceUpload{
+					Name:      meta.Name,
+					Type:      meta.Type.String(),
+					Filename:  uploads[name],
```

Comment:
Filename is from the resource definition in metadata, not what was specified by the user on the CLI. In charm/resource it's Resource.Meta.Path